crates/gitbutler-repo/src/repository_ext.rs

```diff
@@ -156,73 +157,67 @@ impl RepositoryExt for git2::Repository {
-                        } else {
-                            gix::object::tree::EntryKind::Blob
-                        };
-                        (repo.write_blob_stream(&mut file)?, kind)
-                    };
-
-                    head_tree_editor.upsert(rela_path, kind, id)?;
-                }
-                gix::status::index_worktree::iter::Item::Rewrite { .. } => {
-                    unreachable!("disabled")
-                }
-                _ => {}
+        let mut tree_update_builder = git2::build::TreeUpdateBuilder::new();
+
+        let worktree_path = self.worktree_path();
+
+        let statuses = self.statuses(Some(
+            StatusOptions::new()
+                .renames_from_rewrites(false)
+                .renames_head_to_index(false)
+                .renames_index_to_workdir(false)
+                .include_untracked(true)
+                .recurse_untracked_dirs(true),
+        ))?;
+
+        // Truth table for upsert/remove:
+        // | HEAD Tree -> Index | Index -> Worktree | Action    |
+        // | add                | delete            | no-action |
+        // | modify             | delete            | remove    |
+        // |                    | delete            | remove    |
+        // | delete             |                   | remove    |
+        // | delete             | add               | upsert    |
+        // | add                |                   | upsert    |
+        // |                    | add               | upsert    |
+        // | add                | modify            | upsert    |
+        // | modify             | modify            | upsert    |
+
+        for status_entry in &statuses {
+            let status = status_entry.status();
+            let path = status_entry.path().context("Failed to get path")?;
+            let path = Path::new(path);
+
+            if status.is_index_new() && status.is_wt_deleted() {
+                // This is a no-op
+            } else if (status.is_index_deleted() && !status.is_wt_new()) || status.is_wt_deleted() {
+                tree_update_builder.remove(path);
+            } else {
+                let file_path = worktree_path.join(path);
+                let file = std::fs::read(file_path)?;
+                let blob = self.blob(&file)?;
+
+                tree_update_builder.upsert(path, blob, git2::FileMode::Blob);
             }
         }
 
-        let oid = git2::Oid::from_bytes(head_tree_editor.write()?.as_bytes())?;
-        self.find_tree(oid).map(Into::into).map_err(Into::into)
+        let head_tree = self.head_commit()?.tree()?;
+        let tree_oid = tree_update_builder.create_updated(self, &head_tree)?;
+
+        Ok(self.find_tree(tree_oid)?)
+    }
+
+    /// Returns the path of the working directory of the repository.
+    fn worktree_path(&self) -> PathBuf {
+        if self.is_bare() {
+            self.path().to_owned()
+        } else {
+            self.path().join("..")
```

Comment:
This is not correct - the git repository can't be assumed to be a `.git` directory in the worktree.
Instead, [workdir()](https://docs.rs/git2/latest/git2/struct.Repository.html#method.workdir) should be used for that purpose, and it's fine to fail if there is none.
Lastly, if the repository is bare, we wouldn't want to treat the content of a git repository as worktree.